tools/vscode_settings.py

```diff
@@ -1,20 +1,65 @@
 #!/usr/bin/env python3
 
-import json
 from pathlib import Path
 
+try:
+    # VS Code settings allow comments and trailing commas, which are not valid JSON.
+    import json5 as json  # type: ignore[import]
+
+    HAS_JSON5 = True
+except ImportError:
+    import json  # type: ignore[no-redef]
+
+    HAS_JSON5 = False
+
+
+ROOT_FOLDER = Path(__file__).absolute().parent.parent
+VSCODE_FOLDER = ROOT_FOLDER / ".vscode"
+RECOMMENDED_SETTINGS = VSCODE_FOLDER / "settings_recommended.json"
+SETTINGS = VSCODE_FOLDER / "settings.json"
+
+
+# settings can be nested, so we need to recursively update the settings.
+def deep_update(d: dict, u: dict) -> dict:  # type: ignore[type-arg]
+    for k, v in u.items():
+        if isinstance(v, dict):
+            d[k] = deep_update(d.get(k, {}), v)
+        elif isinstance(v, list):
+            d[k] = d.get(k, []) + v
+        else:
+            d[k] = v
+    return d
+
 
 def main() -> None:
-    folder = Path(".vscode")
-    recommended = json.loads((folder / "settings_recommended.json").read_text())
-    path = folder / "settings.json"
+    recommended_setting = json.loads(RECOMMENDED_SETTINGS.read_text())
+    try:
+        current_settings_text = SETTINGS.read_text()
+    except FileNotFoundError:
+        current_settings_text = "{}"
+
     try:
-        current = json.loads(path.read_text())
-    except Exception:
-        current = {}
-    with open(path, "w") as f:
-        json.dump({**current, **recommended}, f, indent=2)
-        f.write("\n")
+        current_settings = json.loads(current_settings_text)
+    except ValueError:  # json.JSONDecodeError is a subclass of ValueError
+        if HAS_JSON5:
+            raise SystemExit("Failed to parse .vscode/settings.json.")
```

Comment:
Why SystemExit and not RuntimeError?

Reply:
Since `tools/vscode_settings.py` is a script to run in the shell. It is not meant to be imported, so its functions will not be put in `try-execpt` blocks in other scripts.

Raising `SystemExit` is equivalent to print an error message and exit with code 1:

```console
$ pip3 uninstall json5
$ ./tools/vscode_settings.py
Failed to parse .vscode/settings.json. Maybe it contains comments or trailing commas. Try `pip install json5` to install an extended JSON parser.
$ echo $?
1
```

Reply:
Ah, thank you for explaining!